.github/workflows/deploy-data.yml

```diff
@@ -0,0 +1,38 @@
+name: Validate/Deploy Data
+
+on:
+  push: { branches: [master] }
+
+jobs:
+  validate-then-deploy:
+    name: Validate, then deploy data
+    runs-on: ubuntu-latest
+    env: task=GH_PAGES
+    steps:
+      - uses: actions/set-up-node@master
+        with: {node-version: ^12.0}
+      - name: Check out the code
+        uses: actions/checkout@master
+      - name: Install dependencies
+        run: yarn install --frozen-lockfile
+      - name: Validate the data
+        run: |
+          npm run validate-bus-data
+          npm run validate-data
+      # Having validated the data, we now prepare a bundle.  This script creates
+      # files in a docs/ directory.
+      - name: Bundle the data
+        run: npm run bundle-data
+      # Notes: actions/checkout@v2 no longer fetches entire history nor enters
+      # detached HEAD state.  We really just need to check out an (orphaned)
+      # branch and then add, commit, and push the appropriate directory.
+      - name: Commit the data
+        run: |
+          git checkout --orphan "gh-pages-$GITHUB_SHA"
+          git --work-tree=docs add .
+          git commit --author="Data Deployment Bot <>" -m "Automated data deployment at $(date -Is)"
+          git show --stat HEAD
+      # If the previous commit successfully happened
+      - name: Deploy the data
+        run:
+          git push --force-with-lease "gh-pages-$GITHUB_SHA:gh-pages"
```

Comment:
Is there a reason to have `gh-pages` twice in the branch name here? I notice above that it's only listed once?

```sh
git checkout --orphan "gh-pages-$GITHUB_SHA"
```

Reply:
Yes, this is one of the advanced features of the `refspec` in `git push <remote> <refspec>`.

The man page for `git-push` says,

```
  <refspec>...
      Specify what destination ref to update with what source object. The
      format of a <refspec> parameter is an optional plus +, followed by the
      source object <src>, followed by a colon :, followed by the destination
      ref <dst>.

      The <src> is often the name of the branch you would want to push, but
      it can be any arbitrary "SHA-1 expression", such as master~4 or HEAD
      (see gitrevisions(7)).

      The <dst> tells which ref on the remote side is updated with this push.
      Arbitrary expressions cannot be used here, an actual ref must be named.
      If git push [<repository>] without any <refspec> argument is set to
      update some ref at the destination with <src> with
      remote.<repository>.push configuration variable, :<dst> part can be
      omitted--such a push will update a ref that <src> normally updates
      without any <refspec> on the command line. Otherwise, missing :<dst>
      means to update the same ref as the <src>.
```

The key here is that typical pushes, `git push origin gh-pages` instruct `git` to update `gh-pages` _on the remote_ with the contents of `gh_pages` _locally_.  The `:` effectively defines a mapping between local ref and remote ref.

Here I'm using `gh-pages-$GITHUB_SHA` as a throwaway branch, but I want to push its contents to the ref `gh-pages` on the remote.  That's why I need `--force-with-lease`, and that's also the reason for the `:`.  Otherwise I'd have to rename or just not use the naming scheme, which I suppose I could do.

(Side note: this is why if you `git push origin :<branch>`, you delete the remote branch; there is no `<src>` so you're overwriting `<branch>` with nothing.)

(Another side note: you can also do cool things like `git push origin :` to push all branches that have a remote-side match, which is helpful for dumping lots of changes at once.)

Reply:
Wow, thanks for explaining that!